mesonbuild/modules/_qt.py

```diff
@@ -269,8 +284,9 @@ def has_tools(self, state: 'ModuleState', args: T.Tuple, kwargs: 'HasToolKwArgs'
             mlog.log('qt.has_tools skipped: feature', mlog.bold(feature), 'disabled')
             return False
         self._detect_tools(state, method, required=False)
-        for tool in self.tools.values():
-            if not tool.found():
+        for tool in kwargs['tools']:
+            assert tool in self.tools, 'tools must be in {moc, uic, rcc, lrelease}'
```

Comment:
This is the third time that same set of tools (`moc, uic, rcc, lrelease`) is explicitly listed out in the code, hardcoded — that's a lot of redundancy and places to keep in sync if the list ever changes. Couldn't it be defined somewhere as a set variable, then used...

* on line 273:

```python
  default=list(_set_of_qt_tools)
```
* on line 274:
 
```python
QtBaseModule._list_in_set_validator(_set_of_qt_tools)
```

* and here:
 
```python
assert tool in self.tools, f'tools must be in {_set_of_qt_tools}'
```

to keep things from potentially getting out of sync?

(Or, heck, at least here as...)

```python
assert tool in self.tools, f'tools must be in {self.tools}'
```

Reply:
fixed, though it doesn't seems that I can reference _set_of_qt_tools for the typechecker in `HasToolKwArgs`

Reply:
Mm, it seems the arg is typed as a `T.List[str]`... perhaps that should be more generic.

Hmm. _Is_ there a generic that covers both `list` and `set`? Neither `Sequence` nor `Mapping` will match both. (I suppose you could call `HasToolKwArgs(list(_set_of_qt_tools))` but that's kind of ugh.)

Reply:
patch has been moved to #13632 (see discussions in #13524). 

> Hmm. Is there a generic that covers both list and set

I don't know about this

Reply:
resolving conversation -- the patch is merged there and should be dropped from this PR in the next rebase.